PowerPC: Add SPE floating point min/max instructions (efsmax, efsmin)

daskey opened the pull request:
#6863 (partial)

Adds support for SPE floating-point min/max instructions (from e200z759CRM):
- efsmax: Floating-Point Single-Precision Maximum  
- efsmin: Floating-Point Single-Precision Minimum

Example:

Assembly:
  efsmin   r5,r3,r4         # r5 = min(r3, r4)
  efsmax   r5,r3,r4         # r5 = max(r3, r4)

Decompiles to:
  // efsmin
  result = fA;
  if (fB < fA) {
      result = fB;
  }

  // efsmax
  result = fA;
  if (fA < fB) {
      result = fB;
  }

daskey committed:
PowerPC: Add SPE floating point min/max instructions (efsmax, efsmin)

GhidorahRex commented on Ghidra/Processors/PowerPC/data/languages/Scalar_SPFP.sinc:
```diff
@@ -467,3 +467,40 @@ define pcodeop ConvertFloatingPointFromUnsignedFraction;
+:efsmin D,A,B is OP=4 & D & A & B & XOP_0_10=0x2B1
+{
+   local result:4;
+   if ( A:4 f<= B:4 ) goto <use_a>;
```
Similarly here

```
   local cond:1 = A:4 f<= B:4;
   result = (zext(cond) * A:4) + (zext(!cond) * B:4);
```

daskey committed:
Refactor efsmax and efsmin for clarity and efficiency

daskey commented on Ghidra/Processors/PowerPC/data/languages/Scalar_SPFP.sinc:
```diff
@@ -467,3 +467,40 @@ define pcodeop ConvertFloatingPointFromUnsignedFraction;
+:efsmax D,A,B is OP=4 & D & A & B & XOP_0_10=0x2B0
+{
+   local result:4;
+   if ( A:4 f>= B:4 ) goto <use_a>;
```
Any luck finding a workaround or are we okay to use gotos in this case? I had a look at similar instructions for other processors, most are using goto

Copilot committed:
Initial plan

Copilot committed:
PowerPC: Use goto pattern for efsmax/efsmin floating-point min/max

Switch from conditional assignment (zext(cond) * inA + zext(!cond) * inB) to
goto-based conditional with local temporaries for cleaner decompiler output.

Co-authored-by: daskey <13271393+daskey@users.noreply.github.com>

daskey committed:
Merge pull request #3 from daskey/copilot/fix-wiki-page-issue

PowerPC: Use goto pattern for efsmax/efsmin floating-point min/max
